File: python3/Trees/balanced_binary_tree_validation.py
from ds import TreeNode

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
"""


def balanced_binary_tree_validation(root: TreeNode) -> bool:
    return get_height_imbalance(root) != -1

def get_height_imbalance(node: TreeNode) -> int:
    # Base case: if the node is null, its height is 0.
    if not node:
        return 0
    # Recursively get the height of the left and right subtrees. If
    # either subtree is imbalanced, propagate -1 up the tree.
    left_height = get_height_imbalance(node.left)
    right_height = get_height_imbalance(node.right)
    if left_height == -1 or right_height == -1:
        return -1
    # If the current node's subtree is imbalanced
    # (height difference > 1), return -1.
    if abs(left_height - right_height) > 1:
        return -1
    # Return the height of the current subtree.
    return 1 + max(left_height, right_height)

File: cpp/Trees/balanced_binary_tree_validation.cpp
#include "ds/TreeNode.h"
using ds::TreeNode;

/**
 * Definition of TreeNode:
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

bool balancedBinaryTreeValidation(TreeNode* root) {
    return getHeightImbalance(root) != -1;
}

int getHeightImbalance(TreeNode* node) {
    // Base case: if the node is null, its height is 0.
    if (!node)
        return 0;
    // Recursively get the height of the left and right subtrees. If
    // either subtree is imbalanced, propagate -1 up the tree.
    int leftHeight = getHeightImbalance(node->left);
    int rightHeight = getHeightImbalance(node->right);
    if (leftHeight == -1 || rightHeight == -1)
        return -1;
    // If the current node's subtree is imbalanced
    // (height difference > 1), return -1.
    if (abs(leftHeight - rightHeight) > 1)
        return -1;
    // Return the height of the current subtree.
    return 1 + std::max(leftHeight, rightHeight);
}